zingo-memo/src/memo_serde.rs

```diff
@@ -0,0 +1,156 @@
+use std::io;
+
+use crate::{
+    utils::{
+        read_transaction_relative_height_and_index, read_unified_address_from_raw_encoding,
+        recover_absolute_transaction_heights, write_transaction_height_and_index,
+        write_unified_address_to_raw_encoding,
+    },
+    ParsedMemo,
+};
+use zcash_client_backend::address::UnifiedAddress;
+use zcash_encoding::{CompactSize, Vector};
+use zcash_primitives::consensus::BlockHeight;
+
+/// Packs a list of UAs into a memo. The UA only memo is version 0 of the protocol
+/// Note that a UA's raw representation is 1 byte for length, +21 for a T-receiver,
+/// +44 for a Sapling receiver, and +44 for an Orchard receiver. This totals a maximum
+/// of 110 bytes per UA, and attempting to write more than 510 bytes will cause an error.
+pub fn create_memo_v0(uas: impl AsRef<[UnifiedAddress]>) -> io::Result<[u8; 511]> {
+    let mut uas_bytes_vec = Vec::new();
+    CompactSize::write(&mut uas_bytes_vec, 0usize)?;
+    Vector::write(&mut uas_bytes_vec, uas.as_ref(), |mut w, ua| {
+        write_unified_address_to_raw_encoding(&ua, &mut w)
+    })?;
+    let mut uas_bytes = [0u8; 511];
+    if uas_bytes_vec.len() > 511 {
+        Err(io::Error::new(
+            io::ErrorKind::InvalidData,
+            "Too many uas to fit in memo field",
+        ))
+    } else {
+        uas_bytes[..uas_bytes_vec.len()].copy_from_slice(uas_bytes_vec.as_slice());
+        Ok(uas_bytes)
+    }
+}
+
+pub fn create_memo_v1(
+    uas: impl AsRef<[UnifiedAddress]>,
+    mut transaction_heights_and_indexes: Vec<(BlockHeight, usize)>,
```

Comment:
Can't we force this to be a u16 or u32..   what's the biggest txindex?

Reply:
`CompactSize::write` takes a `usize` as its argument, so we either take a usize ourselves, or take something else and cast/convert.